### ISSUE TITLE
Build runtime package with required libs

### ISSUE THREAD

**Issue by plata** [PhoenicisOrg/phoenicis-winebuild#62]
Some applications require additional libraries. As discussed, these shall be provided as part of the Wine package.

We need (each 32 and 64 bit):
- openal (LoL)
- ldap (ElsterFormular)

List to be continued.

**Comment by ImperatorS79** [comment]
Here is a list computed from running ldd on every .dll.so:
[libs.txt](https://github.com/PhoenicisOrg/phoenicis-winebuild/files/2825464/libs.txt)

But it is missing the deps of the deps ^^.

But I do not think it should be added inside every .tar.gz, but be downloaded once and installed once (and then use LD_LIBRARY_PATH).

If the file is really big, one could imagine a system that only downloads the libs that are not installed on the user system (using `ldconfig -p | grep lib.so.x` for example).

We should also try to force the user to install basic i386 on his system in the .deb (using control file ?).

**Comment by ImperatorS79** [comment]
Ok, here is a little bash script that do the job for the libs.txt I provided:
```bash
#!/bin/bash

if [[ -z "$1" ]]; then
	echo "Please pass the lib file as argument"
	exit 1
fi

LIBFILE=$1

while read FILELINE; do
	if [[ $FILELINE != libwine.so.* ]]; then
	    echo "Checking for $FILELINE."
	    EXISTS=$(ldconfig -p | grep $FILELINE)
	    if [[ "$EXISTS" != *"$FILELINE"* ]]; then
	    	echo "$FILELINE is not installed on the system!"
	    	FINDINGS=$(apt-file find $FILELINE)
	    	if [[ -z "$FINDINGS" ]]; then
	    		echo "Seems no .deb provides $FILELINE."
	    	else
	    		FINDINGSNAME=${FINDINGS%:*}
	    		echo ".deb $FINDINGSNAME provides $FILELINE."
	    	fi
	    else
	    	echo "$FILELINE is already installed!"
	    fi
	fi
done < $LIBFILE
```

Meaningful output:
```
Checking for libavcodec.so.57.
libavcodec.so.57 is not installed on the system!
Seems no .deb provides libavcodec.so.57.
Checking for libavutil.so.55.
libavutil.so.55 is not installed on the system!
Seems no .deb provides libavutil.so.55.
Checking for libm.so.6.
libm.so.6 is already installed!
Checking for libc.so.6.
libc.so.6 is already installed!
Checking for libz.so.1.
libz.so.1 is already installed!
Checking for libvkd3d.so.1.
libvkd3d.so.1 is already installed!
...
Checking for libpulsecommon-12.2.so.
libpulsecommon-12.2.so is not installed on the system!
.deb libpulse0 provides libpulsecommon-12.2.so.
Checking for libdbus-1.so.3.
libdbus-1.so.3 is already installed!
...
```

**Comment by ImperatorS79** [comment]
But is it so much difficult to do ? I do not think so.

**Comment by plata** [comment]
I'm not sure. For example: when would they be deleted/installed?

**Comment by ImperatorS79** [comment]
Three case:
- Flatpak user does not seem to have missing dependencies, as far as I understand.
- (potentialy) distribution provided .deb user should have already a lot of dependencies installed (control file) (for example I am only missing two libs).
- .deb provided by phoenicis atm, which does not force you to install any deps (even i386).

So running the "script" only once should be enough. You could ask the user: those libs can be installed through packaging system, those libs cannot. And maybe add a button somewhere if the user wants to run again the "script" (after updating for example).

**Comment by plata** [comment]
The idea was to have a general set of libs for everybody (also flatpak and OSX).

**Comment by ImperatorS79** [comment]
OSX users does not ave this problem since the wine package already embed the libs if I understand correctly (maybe it could also be a good idea to put them in a dedicated folder too on OSX to avoid duplication). For flatplak I do not know, you said before that there was no problem.

For me, the idea is to give the user the required libs when their is a version mismatch wit system one, without downloading libs that are already installed (the whole set of wine deps is 700 Mo minimum on my system, it is completely nightmare to put them inside each wine package, and to download all of them when most of them are already installed on the system).

**Comment by Gcenx** [comment]
> OSX users does not ave this problem since the wine package already embed the libs if I understand correctly (maybe it could also be a good idea to put them in a dedicated folder too on OSX to avoid duplication). For flatplak I do not know, you said before that there was no problem.

For macOS the libs are indeed embedded for 32bit builds it only includes 32bit versions to save space, for 64bit it keep the full universal file and now symlinks into /lib64 to ensure wine64 works.

The problem for macOS trying to not include the libs within the wine engine means needing to work around SIP, SIP strips _DYLD_FALLBACK_LIBRARY_PATH_ meaning then needing to workaround that in fun ways. If the intention is really to shrink the size then yes removing them could help and the better workaround for SIP would be making symlinks for all libs into /lib and /lib64.

**Comment by plata** [comment]
@Gcenx the idea really was to provide the required libs and avoid dependency hell. So basically something like we have for OSX just more general also for Linux.

@qparis can you please join the discussion as well? I think we need your input.

**Comment by qparis** [comment]
@Gcenx SIP does not strip _DYLD_FALLBACK_LIBRARY_PATH_. In fact, only system process strips them (including bash), so it is easy to workaround. 
700Mo is in fact very large. Are the lib striped?

**Comment by Gcenx** [comment]
@plata that sounds good since the major grip for PoL seems to be dependancies been an issue.

@qparis yes sorry I forgot this project does not use bash scripts 
Wineskin does for custom wine executable names plus workarou day and winebottler also use bash scripts for similar reasons.
But yeah 700mb seems a lot....

**Comment by plata** [comment]
I would really be interested where these 700mb are coming from.

**Comment by qparis** [comment]
@GCenx, POLv4 use bash script (a lot in fact).
SIP does not prevent you from setting DYLD_FALLBACK_LIBRARY_PATH when you run wine, it prevents you from setting SET_DYLD_FALLBACK_LIBRARY_PATH when you run bash. 
It means that a bash script _can_ export DYLD_FALLBACK_LIBRARY_PATH, but the variable will be strip if bash runs _another_ bash script 

As a workaround, you can chose an arbitrary variable name, like SET_DYLD_FALLBACK_LIBRARY_PATH, and run DYLD_FALLBACK_LIBRARY_PATH="$SET_DYLD_FALLBACK_LIBRARY_PATH" before running wine.

**Comment by Gcenx** [comment]
@qparis I don’t use POL so I wasn’t sure, but yes we do something similar also having 
DYLD_FALLBACK_LIBRARY_PATH=“${WINESKIN_LIB_PATH_FOR_FALLBACK}” that’s in both wine/wine64 or wine-preloaded/wine64-preloader as required including setting freetype to use 35 rendering so we only need to provide the masters freetype version but render fonts correctly on older wine versions.

**Comment by plata** [comment]
I would suggest to add the libraries we know are required and then see where we end up with it. If it becomes to big, we have to think again.

**Comment by Kreyren** [comment]
> * openal (LoL)

Why do we need openal for lol? -> If the issue is only with openal we can exclude it since it's not needed for lol afaik.

**Comment by ImperatorS79** [comment]
@Kreyren This is a general issue, if openal is a dependency of wine, and cannot be installed from the system with the right version, then it has to be downloaded in the "runtime".

Concerning League of Legends, it is a FACT: when you launch the installer and look at the log, you clearly see that the hang is due to missing libavcodec (at the version wine is asking).

**Comment by Kreyren** [comment]
@ImperatorS79 League Of Legends works without openal afaik my GPU AMD7870 doesn't support OpenAL anyway.. It may require it because of something else missing never notice it on my system.

**Comment by ImperatorS79** [comment]
OpenAL have nothing to do with the GPU. It is OpenAudioLibrary.

I do not know for the game itself, but the **installer** needs libavcodec.

**Comment by Kreyren** [comment]
@ImperatorS79 Noted, still doubt that it's needed afaik i'm running leagueoflegends atm without it.. Not even lutris mensiones it on https://lutris.net/games/install/3552/view

About `msvcp140` works without this on my system afaik.

**Comment by plata** [comment]
@qparis you mentioned that you wanted to introduce a runtime. Can you elaborate how this should look like?

**Comment by qparis** [comment]
I think it will just ba a python file that extract libraries from Docker image. Then a unique name will be given and we host it.
Does it sound ok?

**Comment by ImperatorS79** [comment]
How will phoenicis download that runtime? Will phoenicis be able to only download the libs that are not installed ?

**Comment by qparis** [comment]
In a first time, we could download the whole archive, right?

**Comment by ImperatorS79** [comment]
Of course, but it just a frienldy reminder that it should not be the perfect solutions ;-)

**Comment by qparis** [comment]
Do you mean that we do not need a perfect solution for the moment?

**Comment by ImperatorS79** [comment]
I mean if you want to create a first implementation that just download to whole runtime to show that it solves the problem, you can of course do it.

But do not close this issue (or create a new one) for the part where the whole runtime is not downloaded.

**Comment by plata** [comment]
When will the archive be downloaded? Every time you install an application?

**Comment by qparis** [comment]
Same rules than wine I guess. We

**Comment by plata** [comment]
Does it mean it's something which must be called explicitly in the scripts?

**Comment by qparis** [comment]
No, we can make wine installation do the stuff

**Comment by plata** [comment]
Good. I think the approach is good. It offers a lot of flexibility. Do you want to implement a prototype?

**Comment by qparis** [comment]
I'm currently working on a prototype to expose the runtimes on a webservice

**Comment by qparis** [comment]
Or at least a static file bucket for the moment

**Comment by plata** [comment]
Yeah, that should be sufficient for now.

**Comment by plata** [comment]
As you said runtimes: Is there more than one?

**Comment by qparis** [comment]
32bits, 64bits, linux, osx

**Comment by plata** [comment]
Ok like this. I started to fear a runtime per app.

**Comment by qparis** [comment]
Can someone test this runtime? https://www.playonlinux.com/runtime.tar.gz

**Comment by Kreyren** [comment]
@qparis intructions?

**Comment by qparis** [comment]
Extract it and export LD_LIBRARY_PATH to lib/ folder before running phoenicis

**Comment by Kreyren** [comment]
@plata Test using league of legends commit sufficient?

**Comment by qparis** [comment]
Should be a good start yes

**Comment by Kreyren** [comment]
results in: 

![image](https://user-images.githubusercontent.com/11302521/53687537-14eb3d00-3d36-11e9-8964-7a60f9d705a8.png)

```
java.lang.ClassCastException: class jdk.nashorn.api.scripting.ScriptObjectMirror cannot be cast to class java.lang.String (jdk.nashorn.api.scripting.ScriptObjectMirror is in module jdk.scripting.nashorn of loader 'platform'; java.lang.String is in module java.base of loader 'bootstrap')
	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$9$\^eval\_.:scopeCall(<eval>)
	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$9$\^eval\_.:program(<eval>:5)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:655)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:513)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:527)
	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:456)
	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:413)
	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:409)
	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:162)
	at java.scripting/javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by qparis** [comment]
This is absolutely not related

**Comment by Kreyren** [comment]
@qparis it's league of legends script invoked on phoenicis with `LD_LIBRARY_PATH`.

So it's problem in a script? Should i test something else?

**Comment by qparis** [comment]
I don't know what the problem is but this discussion does not belong to phoenicis-winebuild repository. I think you have a misconfiguration on your side

**Comment by ImperatorS79** [comment]
libavcodec and libavutil are missing from the runtime, which are the one needed to run the league of legends installer.

@qparis How did you compute this runtime ? I think there are missing deps in comparison to the libs.txt file I posted earlier here.

**Comment by qparis** [comment]
Manually, unfortunately because wine is not necesseraly compiled when it is generated

**Comment by qparis** [comment]
does it work if you add libavcodec and libavutil?

**Comment by ImperatorS79** [comment]
Yes the installer will work with those two libs, but the problem of @Kreyren was that he did not update his script for the new include mechanism.

You should add to the runtime the libs from the libs.txt file I posted here.

**Comment by Kreyren** [comment]
@qparis Tested with new installer results in:
![image](https://user-images.githubusercontent.com/11302521/53688660-8b446b00-3d47-11e9-9ee4-00414ecb56f7.png)

```
java.lang.RuntimeException: java.io.IOException: Cannot run program "/home/kreyren/.Phoenicis/engines//wine/staging-linux-x86/4.2/bin/wine" (in directory "/home/kreyren/.Phoenicis/containers/wineprefix/League of Legends"): error=2, No such file or directory
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.linker.JavaAdapterServices.wrapThrowable(JavaAdapterServices.java:285)
	at jdk.nashorn.javaadapters.org_phoenicis_engines_Engine.run(Unknown Source)
	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$178$4947AAAZZA$\^eval\_.run(engines.wine.engine.object:178)
	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$177$219A$\^eval\_$cu1$restOf.regedit#open(engines.wine.plugins.regedit:12)
	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$164$287$\^eval\_$cu1$restOf.luna(engines.wine.verbs.luna:29)
	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$108$493$\^eval\_$cu1$restOf.go(engines.wine.quick_script.installer_script:72)
	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$76$603$\^eval\_.run(<eval>:18)
	at jdk.nashorn.javaadapters.org_phoenicis_scripts_Installer.run(Unknown Source)
	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$75$\^eval\_$cu1$restOf.:program(<eval>:133)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:655)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:513)
	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:527)
	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:456)
	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:413)
	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:409)
	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:162)
	at java.scripting/javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.io.IOException: Cannot run program "/home/kreyren/.Phoenicis/engines//wine/staging-linux-x86/4.2/bin/wine" (in directory "/home/kreyren/.Phoenicis/containers/wineprefix/League of Legends"): error=2, No such file or directory
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1128)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1071)
	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$217$7066AAAZZA$\^eval\_$cu1$restOf.run(engines.wine.engine.implementation:267)
	... 22 more
Caused by: java.io.IOException: error=2, No such file or directory
	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
	at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:340)
	at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:271)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1107)
	... 24 more
```

![image](https://user-images.githubusercontent.com/11302521/53688664-97c8c380-3d47-11e9-8a55-dd6587859241.png)

Note that phoenicis is invoked using http://ix.io/1CpV with prefix for required libraries that are extracted.

Is it relevant now?

**Comment by qparis** [comment]
/home/kreyren/.Phoenicis/engines//wine/staging-linux-x86/4.2/bin/wine does not seem to exist.

**Comment by Kreyren** [comment]
> /home/kreyren/.Phoenicis/engines//wine/staging-linux-x86/4.2/bin/wine does not seem to exist.

kreyren@debian:~$ ls /home/kreyren/.Phoenicis/engines/wine/ | ix
http://ix.io/1CpX

Can confirm.

**Comment by qparis** [comment]
Seems to be a problem on your side...

**Comment by Kreyren** [comment]
@qparis reinstalled phoenicis prior to the experiment and nuked .Phoenicis. 

Checking script

**Comment by qparis** [comment]
> @qparis reinstalled phoenicis prior to the experiment and nuked .Phoenicis.
> 
> Checking script
> 
> EDIT: i don't see any issues in a script.. i may be wrong since i have no effective way to verify it's sanity..
> 
> EDIT2: Unless followed is not sane:
> 
> ```js
> .wineDistribution("staging")
> .wineVersion(LATEST_STAGING_VERSION)
> ```
> 
> EDIT3: maybe should i download required wine in phoenicis manually and try again after?
> 
> EDIT4: I'm not sane those are required, seems unsane to me.. Need verification?
> 
> ```js
> return {command: setupFile, args: ["--installdir", "C:/Riot Games/League of legends/", "--mode", "unattended"]};
> ...
>   mkdir(wine.prefixDirectory() + "drive_c/Riot Games/League of legends/");
> 
>                 Create run script to start the right exe
>                 var client = wine.prefixDirectory() + "drive_c/Riot Games/League of Legends/";
>                 var batContent = "start C:/Riot Games/League of legends/";
>                 writeToFile(client, batContent);
> ```
> 
> EDIT5: Downloaded wine manually and it fixes previous issue, but now hangs on
> ![image](https://user-images.githubusercontent.com/11302521/53689282-9b157c80-3d52-11e9-97dc-d8f2e14c4046.png)
> 
> EDIT6: results in fatal:
> 
> ![image](https://user-images.githubusercontent.com/11302521/53689299-fe9faa00-3d52-11e9-999e-8a978f1ba94f.png)
> 
> ```
> java.lang.RuntimeException: java.io.IOException: Cannot run program "/home/kreyren/.Phoenicis/engines//wine/staging-linux-x86/4.2/bin/wine" (in directory "/home/kreyren/.Phoenicis/containers/wineprefix/League of Legends"): error=2, No such file or directory
> 	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.linker.JavaAdapterServices.wrapThrowable(JavaAdapterServices.java:285)
> 	at jdk.nashorn.javaadapters.org_phoenicis_engines_Engine.run(Unknown Source)
> 	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$833$4947AAAZZA$\^eval\_.run(engines.wine.engine.object:178)
> 	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$832$219A$\^eval\_$cu1$restOf.regedit#open(engines.wine.plugins.regedit:12)
> 	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$819$287$\^eval\_$cu1$restOf.luna(engines.wine.verbs.luna:29)
> 	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$778$493$\^eval\_$cu1$restOf.go(engines.wine.quick_script.installer_script:72)
> 	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$746$603$\^eval\_.run(<eval>:18)
> 	at jdk.nashorn.javaadapters.org_phoenicis_scripts_Installer.run(Unknown Source)
> 	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$745$\^eval\_$cu1$restOf.:program(<eval>:133)
> 	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:655)
> 	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:513)
> 	at jdk.scripting.nashorn/jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:527)
> 	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:456)
> 	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:413)
> 	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:409)
> 	at jdk.scripting.nashorn/jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:162)
> 	at java.scripting/javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
> 	at org.phoenicis.scripts.nashorn.NashornEngine.eval(NashornEngine.java:50)
> 	at org.phoenicis.scripts.nashorn.NashornScriptInterpreter.runScript(NashornScriptInterpreter.java:35)
> 	at org.phoenicis.scripts.interpreter.BackgroundScriptInterpreter.lambda$runScript$0(BackgroundScriptInterpreter.java:35)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> 	at java.base/java.lang.Thread.run(Thread.java:834)
> Caused by: java.io.IOException: Cannot run program "/home/kreyren/.Phoenicis/engines//wine/staging-linux-x86/4.2/bin/wine" (in directory "/home/kreyren/.Phoenicis/containers/wineprefix/League of Legends"): error=2, No such file or directory
> 	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1128)
> 	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1071)
> 	at jdk.scripting.nashorn.scripts/jdk.nashorn.internal.scripts.Script$Recompilation$890$7066AAAZZA$\^eval\_$cu1$restOf.run(engines.wine.engine.implementation:267)
> 	... 22 more
> Caused by: java.io.IOException: error=2, No such file or directory
> 	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
> 	at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:340)
> 	at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:271)
> 	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1107)
> 	... 24 more
> ```

Please, this issue is related to winebuild, this is not the right place to discuss about this problem

**Comment by qparis** [comment]
@ImperatorS79 Runtime are still incomplete (merge request are now easier to improve this :)), but at least, the archives are automatically uploaded and the task is created: https://www.playonlinux.com/wine/runtimes/

Next step: Create a server side json and integrate this to phoenicis

**Comment by plata** [comment]
@qparis if you want to try yourself (even if you have the required libs installed already), you can run ElsterFormular in flatpak. The app is free and you will see immediately if it works (otherwise you get a lot of error windows when you start it). Should be especially helpful when integrating the runtime in Phoenicis.

**Comment by Kreyren** [comment]
sorry if i didn't help :X

**Comment by ImperatorS79** [comment]
@qparis There is no runtime currently uploaded for 64 bits.

**Comment by ImperatorS79** [comment]
@qparis Any news about this ?

**Comment by qparis** [comment]
Hmm, though that the runtime was complete. Any idea what could be messing up?
Can someone confirm that if the runtime is manually installed, problems are solved?

**Comment by plata** [comment]
Cannot check right now. However, the best and easiest check in my opinion is running e.g. LoL or ElsterFormular from flatpak as those will definitely fail if the runtime doesn't work. It also ensures that you do not test libraries installed on your OS anyways by mistake.

**Comment by Kreyren** [comment]
![image](https://user-images.githubusercontent.com/11302521/56456097-78214500-635f-11e9-8db0-96163d7c6768.png)
![image](https://user-images.githubusercontent.com/11302521/56456100-81aaad00-635f-11e9-85de-ae771eac62f1.png)
![image](https://user-images.githubusercontent.com/11302521/56456103-8d966f00-635f-11e9-97bb-f0e8820d6ad5.png)

Trying flatpak now

**Comment by Kreyren** [comment]
grabbed from https://www.flathub.org/apps/details/org.phoenicis.playonlinux ~~and i hate it already~~

```console
kreyren@dreamon ~/Downloads $ sudo flatpak install org.phoenicis.playonlinux.flatpakref
Password:
The remote 'flathub', refered to by 'org.phoenicis.playonlinux' at location https://dl.flathub.org/repo contains additional applications.
Should the remote be kept for future installations? [Y/n]: y
The application org.phoenicis.playonlinux depends on runtimes from:
  https://dl.flathub.org/repo/
Configure this as new remote 'flathub-1' [Y/n]: y
Required runtime for org.phoenicis.playonlinux/x86_64/stable (runtime/org.freedesktop.Platform/x86_64/18.08) found in remotes: flathub

   1) flathub
   2) flathub-1

Which do you want to install (0 to abort)? [0-2]:
Which do you want to install (0 to abort)? [0-2]: 1

org.phoenicis.playonlinux permissions:
    ipc                  network      x11      devices      multiarch
    file access [1]

    [1] xdg-documents, xdg-download


        ID                                          Arch        Branch      Remote       Download
 1. [✓] org.freedesktop.Platform                    x86_64      18.08       flathub      279.1 MB / 293.0 MB
 2. [✓] org.freedesktop.Platform.Locale             x86_64      18.08       flathub      160.9 MB / 315.9 MB
 3. [✓] org.freedesktop.Platform.VAAPI.Intel        x86_64      18.08       flathub        1.8 MB / 1.8 MB
 4. [✓] org.freedesktop.Platform.html5-codecs       x86_64      18.08       flathub        4.8 MB / 4.9 MB
 5. [✓] org.phoenicis.playonlinux                   x86_64      stable      flathub      110.6 MB / 110.7 MB
 6. [✓] org.freedesktop.Platform.Compat.i386        x86_64      18.08       flathub      177.3 MB / 177.8 MB

Installation complete.
```

__UPDATING__

**Comment by Kreyren** [comment]
Suspect that alpha_2 will not work due missing `adobeair`.

**Comment by plata** [comment]
In your first post, the Wine latest version variable has not been resolved. Looks like you maybe put the variable in quotation marks.
For your flatpak test: You cannot use scripts master with alpha-2 software.

**Comment by Kreyren** [comment]
> variable in quotation marks.

yep i did, trying without quotation marks

**Comment by Kreyren** [comment]
updated posts above with results

**Comment by plata** [comment]
Looks like a different issue with LoL to me. I don't think it's related to this topic at all. To be clear: This is about runtime dependencies. If an app does not install, that's something else.

**Comment by Kreyren** [comment]
> Looks like a different issue with LoL to me. I don't think it's related to this topic at all. To be clear: This is about runtime dependencies. If an app does not install, that's something else.

Note that it does install on phoenicis compiled from master -> not script issue?

**Comment by ImperatorS79** [comment]
@qparis If you look at the very beginning of this issue, I tried to solve the libavutil/libavcodec solution with a symlink only solution, and it did not work because wine required version 57 and nothing else. You can download the runtime.tar.gz and see that inside there are mostly symlinks. This will definitely not work.

If the runtime was available for 64 and 32 bits, then I am sure @plata or @madoar could certainly easily handle it in phoenicis.

**Comment by Kreyren** [comment]
@qparis any info? Can't i just provide my own wine at this point?

**Comment by qparis** [comment]
> @qparis any info? Can't i just provide my own wine at this point? No, there are no reasons to believe that your build will work better for others. We absolutely don't need to fragmentate this even more.

Can someone using Linux compile the last 32bits runtime and provides me a status of what is missing so that it works properly?

**Comment by plata** [comment]
@qparis do we have a description on how to compile/use the runtime?

**Comment by qparis** [comment]
Running example/runtime_linux.py with the right python path should be enough

**Comment by plata** [comment]
Any changes required to the Wine script such that the environment is actually used (e.g. `LD_LIBRARY_PATH`)?

Just to ensure that anybody reading this can test. Normally, I would test myself in a fresh Manjaro VM (because I know that some of the dependencies required for ElsterFormular are not installed there by default). However, I cannot run a VM currently.

**Comment by qparis** [comment]
LD_LIBRARY_PATH yes

**Comment by plata** [comment]
Just tried to test without VM. Problem is ElsterFormular still worked after uninstalling ldap. Maybe this dependencies is no longer required after a recent update. As I do not have LoL, I do not know what I could test currently.

@ImperatorS79 can you try?

**Comment by ImperatorS79** [comment]
@plata I do not have time currently. But could you look at the builded .zip to see if the libs are symlink or not ?

**Comment by Kreyren** [comment]
> Can someone using Linux compile the last 32bits runtime and provides me a status of what is missing so that it works properly?

> Running example/runtime_linux.py with the right python path should be enough

Found https://github.com/PhoenicisOrg/phoenicis-winebuild/blob/master/examples/example_linux.py

Provide instructions if testing is needed.

**Comment by ImperatorS79** [comment]
@qparis OK I just tested it. Again libs are 90% symlinks. without the real libs it will not work. Also:
```bash
cp: cannot stat 'libvk3d*.so*': No such file or directory
cp: cannot stat 'libheimbase*.so*': No such file or directory
```

**Comment by ImperatorS79** [comment]
Ok so results from my test with the runtime:

-The runtime is 544 MB uncompressed (amd64 + x64) (ouch).
-Using the full runtime makes LoL installer crash, but using only the two missing libs on my system makes it install successfully.

So as I suggested earlier in this issue, phoenicis should only add in the installed runtime libs that are not present on the user system in order to minimize the risk of crash (and the size btw).

**Comment by plata** [comment]
Any idea why it crashes?

> using only the two missing libs on my system makes it install successfully

Could this be solved by simply putting the runtime after the system library path for `LD_LIBRARY_PATH`?

**Comment by ImperatorS79** [comment]
I did
```
LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:<path_to_runtime> wine League\ of\ Legends\ installer\ EUW.exe 
```
And it crashes.

**Comment by plata** [comment]
Mh. No idea how this can happen. Maybe an issue with the rpath of the wine executable?

**Comment by qparis** [comment]
Any error about the cause of the crash?

**Comment by ImperatorS79** [comment]
Nothing, just the unhandled page fault.

**Comment by qparis** [comment]
With or without the runtime?

**Comment by ImperatorS79** [comment]
With, without the installer crashes when trying to install DirectX9, and with only libavutil libavcodec in the runtime in runs dine

**Comment by qparis** [comment]
So does it mean that we are only missing libavutil and libavcodec in the runtime so that it works?

**Comment by ImperatorS79** [comment]
It means that, on my config, the only libs that are missing for wine are libavcodec and libavutils, and only beacause wine requires a specific version for them. However, I did not test this a lot and other libs might be in the same case, and it is also user dependant. I still have POL4 installed, and so all the wine deps from the wine package. The problem is a user installing phoenicis from the current .deb does not install any of those deps :/.

**Comment by qparis** [comment]
Are these libs included in this runtime?

**Comment by ImperatorS79** [comment]
Yes, but using the full runtime crashes Lol installer at 0 seconds.

**Comment by qparis** [comment]
Can you remove the librairie one by one to see which one causes problems?

**Comment by ImperatorS79** [comment]
Yes I could do that, but how can you be sure that this error will not happen with another lib on someone else computer ?

**Comment by qparis** [comment]
I think it will depends on the guilty library

**Comment by ImperatorS79** [comment]
I managed to have this:
```
Register dump:
 CS:0023 SS:002b DS:002b ES:002b FS:0063 GS:006b
 EIP:00000000 ESP:008bedfc EBP:008bef68 EFLAGS:00010202(  R- --  I   - - - )
 EAX:00000031 EBX:7e59c000 ECX:00000063 EDX:00000001
 ESI:7e578dba EDI:7e5a684c
Stack dump:
0x008bedfc:  7e54aff7 7e018ec0 0013f334 008beec8
0x008bee0c:  7d23b143 7d298000 008befd4 0013ee60
0x008bee1c:  7d23b171 7e018ec0 008bee84 00000001
0x008bee2c:  7e5519f3 7e5a67e0 008bee5e 008beef4
0x008bee3c:  00002000 008b0000 7e5a683c 008bef10
0x008bee4c:  7e5a683c 0013f2e0 008beed4 7d1f1000
Backtrace:
=>0 0x00000000 (0x008bef68)
  1 0x7d26b01b xrenderdrv_SelectFont+0x8a() in winex11 (0x008bf058)
  2 0x7e53566d FONT_SelectObject+0x8c() in gdi32 (0x008bf138)
  3 0x7e552aed SelectObject+0xec() in gdi32 (0x008bf178)
  4 0x0047b260 in league of legends installer euw (+0x7b25f) (0x008bf33c)
  5 0x7e019920 (0x7e52f65f)
0x00000000: -- no code accessible --
Modules:
Module  Address                 Debug info      Name (71 modules)
PE        400000-  6be000       Export          league of legends installer euw
ELF     7b400000-7b800000       Deferred        kernel32<elf>
  \-PE  7b420000-7b800000       \               kernel32
ELF     7bc00000-7bd11000       Deferred        ntdll<elf>
  \-PE  7bc10000-7bd11000       \               ntdll
ELF     7c000000-7c004000       Deferred        <wine-loader>
ELF     7cee3000-7cf1b000       Deferred        uxtheme<elf>
  \-PE  7cef0000-7cf1b000       \               uxtheme
ELF     7cf1b000-7cf22000       Deferred        libxfixes.so.3
ELF     7cf22000-7cf2f000       Deferred        libxcursor.so.1
ELF     7cf2f000-7cf42000       Deferred        libxi.so.6
ELF     7cf42000-7cf4f000       Deferred        libxrandr.so.2
ELF     7cf4f000-7cf5b000       Deferred        libxrender.so.1
ELF     7d05b000-7d07a000       Deferred        libbsd.so.0
ELF     7d07a000-7d0a8000       Deferred        libxcb.so.1
ELF     7d0a8000-7d1f4000       Deferred        libx11.so.6
ELF     7d1f4000-7d209000       Deferred        libxext.so.6
ELF     7d209000-7d2a3000       Dwarf           winex11<elf>
  \-PE  7d210000-7d2a3000       \               winex11
ELF     7d2c8000-7d302000       Deferred        libpng16.so.16
ELF     7d302000-7d31f000       Deferred        libz.so.1
ELF     7d31f000-7d3d3000       Deferred        libfreetype.so.6
ELF     7d3d3000-7d3f8000       Deferred        libtinfo.so.5
ELF     7d3f8000-7d420000       Deferred        libncurses.so.5
ELF     7d441000-7d574000       Deferred        oleaut32<elf>
  \-PE  7d460000-7d574000       \               oleaut32
ELF     7d574000-7d63d000       Deferred        msvcrt<elf>
  \-PE  7d590000-7d63d000       \               msvcrt
ELF     7d63d000-7e006000       Deferred        shell32<elf>
  \-PE  7d650000-7e006000       \               shell32
ELF     7e02a000-7e02e000       Deferred        libxcomposite.so.1
ELF     7e02e000-7e035000       Deferred        libxxf86vm.so.1
ELF     7e035000-7e06f000       Deferred        ws2_32<elf>
  \-PE  7e040000-7e06f000       \               ws2_32
ELF     7e06f000-7e0b1000       Deferred        winspool<elf>
  \-PE  7e080000-7e0b1000       \               winspool
ELF     7e0b1000-7e13e000       Deferred        rpcrt4<elf>
  \-PE  7e0c0000-7e13e000       \               rpcrt4
ELF     7e13e000-7e2a6000       Deferred        ole32<elf>
  \-PE  7e160000-7e2a6000       \               ole32
ELF     7e2a6000-7e2ca000       Deferred        shcore<elf>
  \-PE  7e2b0000-7e2ca000       \               shcore
ELF     7e2ca000-7e33a000       Deferred        shlwapi<elf>
  \-PE  7e2e0000-7e33a000       \               shlwapi
ELF     7e33a000-7e432000       Deferred        comdlg32<elf>
  \-PE  7e340000-7e432000       \               comdlg32
ELF     7e432000-7e458000       Deferred        imm32<elf>
  \-PE  7e440000-7e458000       \               imm32
ELF     7e458000-7e4a4000       Deferred        usp10<elf>
  \-PE  7e460000-7e4a4000       \               usp10
ELF     7e4a4000-7e5f8000       Dwarf           gdi32<elf>
  \-PE  7e4b0000-7e5f8000       \               gdi32
ELF     7e5f8000-7e82a000       Deferred        user32<elf>
  \-PE  7e610000-7e82a000       \               user32
ELF     7e82a000-7e980000       Deferred        comctl32<elf>
  \-PE  7e830000-7e980000       \               comctl32
ELF     7e980000-7e9fb000       Deferred        advapi32<elf>
  \-PE  7e990000-7e9fb000       \               advapi32
ELF     7eec7000-7eedb000       Deferred        libnss_files.so.2
ELF     7eedb000-7efdf000       Deferred        libm.so.6
ELF     7efe1000-7efe6000       Deferred        libxinerama.so.1
ELF     7efe6000-7f000000       Deferred        version<elf>
  \-PE  7eff0000-7f000000       \               version
ELF     f7b02000-f7b0c000       Deferred        librt.so.1
ELF     f7b0e000-f7b14000       Deferred        libdl.so.2
ELF     f7b14000-f7cf4000       Deferred        libc.so.6
ELF     f7cf4000-f7d14000       Deferred        libpthread.so.0
ELF     f7d15000-f7d1c000       Deferred        libxdmcp.so.6
ELF     f7d1c000-f7d20000       Deferred        libxau.so.6
ELF     f7d35000-f7eee000       Dwarf           libwine.so.1
ELF     f7ef0000-f7f1a000       Deferred        ld-linux.so.2
Threads:
process  tid      prio (all id:s are in hex)
00000008 (D) C:\users\Public\Documents\League of Legends installer EUW.exe
        0000002d    2
        0000002c    2
        0000002b    2
        00000009    0 <==
0000000e services.exe
        00000029    0
        00000026    0
        0000001f    0
        0000001a    0
        00000017    0
        00000016    0
        00000015    0
        00000011    0
        0000000f    0
00000010 explorer.exe
        00000025    0
        00000022    0
        0000001e    0
        00000012    0
00000013 winedevice.exe
        0000001b    0
        00000019    0
        00000018    0
        00000014    0
0000001c plugplay.exe
        00000021    0
        00000020    0
        0000001d    0
00000023 winedevice.exe
        0000002a    0
        00000028    0
        00000027    0
        00000024    0
System information:
    Wine build: wine-4.0
    Platform: i386
    Version: Windows 7
    Host system: Linux
    Host version: 5.0.0-13-generic
```

**Comment by ImperatorS79** [comment]
Okay it is libfreetype

**Comment by qparis** [comment]
Do we have the same version?

**Comment by ImperatorS79** [comment]
On my system it is libfreetype.so.6.16.1, in the runtime it is libfreetype.so.6.12.3

**Comment by qparis** [comment]
Both 32bits?

**Comment by ImperatorS79** [comment]
I have 32 bits and 64 bits libs installed.

**Comment by qparis** [comment]
I suggest we try to remove freetype from the runtime on linux, and see if it is ok for everyone

**Comment by Kreyren** [comment]
We could also make some method that allows scripts to download required dependencies on demand from predefined source like it was discussed before to permanently resolve issues alike since runtime libs are always required.

Or depending on package manager present on the system we can parse the instructions to it so that those deps would be fetched using distro repo.

**Comment by ImperatorS79** [comment]
@Keyren this can be second phase. First let's implement the runtime since it is working now.

**Comment by Kreyren** [comment]
i personally hate runtimes since they are usually bloated af making system slower.. (based on results on my gentoo where i have libs compiled specifically on the hardware)

**Comment by qparis** [comment]
@kreyren Please stop contesting each decision..

**Comment by Kreyren** [comment]
@qparis You need to realize that this is not exclusive MacOS community.. We in linux blame each other about provided code, decisions and info for Quality assurance which you should welcome..

**Comment by qparis** [comment]
> @qparis You need to realize that this is not exclusive MacOS community.. We in linux blame each other about provided code, decisions and info for Quality assurance which you should welcome..

Please put into practice what you are saying, especially for Gentoo. While you are always pushing for specific distribution solutions, we are trying to imagine agnostic solutions that will not require a different implementation for each configuration and without reinventing the wheel. 

> Probably if you would respect GNU ethics from time to time we would have working phoenicis now 🤔

Please don't make me wrong, code and solution challenging is very welcome. What you are generally proposing are in general soilless concepts, without any working proof of concept behind.
We spend a lot time (probably hours!) to explain to you why the solution is not working, and most of the time, we loop and rexplain the same things. ==> You cannot argue the fact that we are not open to discussion. 
As you said, just before we are lacking time, so I suggest we are efficient on the decision making process. 

I have no vanity for this project, so if you manage to build your own community from scratch with a _better GNU ethics_ , please do so and I will sincerely salute it. 

Because we do not (collectively) agree with one of you solution does not mean that we are not open to the discussion. Thoughts for food. 

Can we now focus back on the discussion? @ImperatorS79 what do you think about making freetype a disitribution dependency and managing the rest within the runtime?

**Comment by ImperatorS79** [comment]
@qparis As I said, this is a good first solution. You also know that I would prefer having the less things possible in the runtime. 

Do you think it would be possible to have a file listing what libs needs to be downloaded from the runtime per distribution? So that we can say to the user : "please installs the following package to ensure wine will work fine". If the distribution is not found in it, you just download the full runtime.

**Comment by qparis** [comment]
The most tricky part will be to detect what librairie you need. How you would handle it?

**Comment by ImperatorS79** [comment]
See the beginning of this issue. I showed to .sh file. One simply run ldd on each wine .so file to determine the libs, the other one use apt-file (this changes depending on the distribution) to determine if the lib exists in the good version in a .deb.

**Comment by qparis** [comment]
apt-file is not agnostic. You'll probably end up with the same problems. The version is not neceserally the only compatibility criteria. 
Note that the runtime will generally be downloaded once

**Comment by ImperatorS79** [comment]
What do you mean by "not agnostic" ?

**Comment by qparis** [comment]
You need to define a different implementation for each package manager

**Comment by ImperatorS79** [comment]
Yes, that what I said in "this changes depending on the distribution". Anyway, I was thinking of a file stored server side. So that, for, a distribution that has been tested and where you are sure that system libs works, prefer system libs from runtime ones.

**Comment by qparis** [comment]
I see. So what you are suggesting is just some kind of "lazy downloading" system. I propose we try to handle it in a second step

**Comment by ImperatorS79** [comment]
Yes, sure.

**Comment by Kreyren** [comment]
> @qparis What you are generally proposing are in general soilless concepts, without any working proof of concept behind. We spend a lot time (probably hours!) to explain to you why the solution is not working, and most of the time, we loop and rexplain the same things. ==> You cannot argue the fact that we are not open to discussion.

Note that i've provided proof of concept to all my proposed solutions that are based on my personal experience. I would never propose something with which is not verified prior to the suggestion. 

>  so I suggest we are efficient on the decision making process.

Efficient maybe, but it feels to me like you are hotfixing everything instead of making permanent solutions with required research..

 There **SHOULD** be someone at lead saying "Ok, we have problem with implementation of XZ for which we need to resovle bug0000 and based on all available informations we decided that doing ABC based on EDF provided is the most effective so development will be focused only on this one function to be implemented now." but instead it seems that you are working on milion stuff without any workable results..

> I have no vanity for this project, so if you manage to build a new community from scratch with better GNU ethics, please do so and I will sincerely salute it.

i have a project that has functions to manage compatibility layers but it is not it's expected usecase and i want to rather use phoenicis since i'm hoping that you make it work.

Phoenicis by itself is very close to gold.. Meaning that if you fix wine issues this project can be pushed directly into gold assuming that it's core is already supperior to lutris. (excluding long download bug and lack of customization of engine using GUI)

**Comment by qparis** [comment]
@Kreyren Please keep this thread available for the discussion about the runtime issue

**Comment by ImperatorS79** [comment]
@qparis Now you should be able to upload amd64 and x86 runtime somewhere, then someone could implement it's usage in phoenicis.

**Comment by qparis** [comment]
I'm working on it right now

**Comment by qparis** [comment]
https://phoenicis.playonlinux.com/index.php/runtime?os=linux

AMD64 is coming

**Comment by plata** [comment]
When should we download the runtime? Best would probably be with every engine installation, right? Shall we add a property to (not) use the runtime?

**Comment by qparis** [comment]
The runtime should be indeed downloaded with every engine installation but only if it has not been installed before

**Comment by plata** [comment]
Is it enough to check if it has not been installed before? What I'm thinking about it: The runtime could change. So either we have to compare a hash or redownload it unconditionally.

**Comment by qparis** [comment]
In a first version it will be sufficient. Later, we may assign a runtime per version

**Comment by ImperatorS79** [comment]
Don't do a runtime per version, runtime is about 500 MB! And since all wine versions are built on the same system, it can be common to all wine versions.

Checking a hash does seem to be a good idea to see if the runtime has changed.

**Comment by qparis** [comment]
A runtime per version of course do not mean a different runtime per version

**Comment by ImperatorS79** [comment]
I guess this is solved.
